snap/integrity/integrity.go

```diff
@@ -0,0 +1,262 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package integrity
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap/integrity/dm_verity"
+	"github.com/snapcore/snapd/snap/squashfs"
+	"golang.org/x/crypto/sha3"
+)
+
+const (
+	blockSize = 4096
+	// https://github.com/plougher/squashfs-tools/blob/master/squashfs-tools/squashfs_fs.h#L289
+	squashfsSuperblockBytesUsedOffset = 40
+)
+
+var (
+	// magic is the magic prefix of snap metadata blocks.
+	magic = []byte{'s', 'n', 'a', 'p', 'e', 'x', 't'}
+)
+
+// Align input `size` to closest `blockSize` value
+func align(size uint64) uint64 {
+	return (size + blockSize - 1) / blockSize * blockSize
+}
+
+type IntegrityData struct {
+	Header         *IntegrityDataHeader
+	Offset         uint64
+	SourceFilePath string
+
+	Bytes *[]byte
```

Comment:
we probably don't want to bring all the data into memory even it's relatively smaller than the full snap, it might still be quite a bit of pressure on small devices

Reply:
Changed `SHA3_384()` to use `osutil.PartialFileDigest()` (and calculate the hash at the time of validation).